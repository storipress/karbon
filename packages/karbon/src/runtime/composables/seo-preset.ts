import { filter, first, identity, isTruthy, map, pathOr, pipe } from 'remeda'
import type { PartialDeep } from 'type-fest'
import type { MetaFlatInput } from '@zhead/schema'
import { isDefined } from '@vueuse/core'
import truncate from 'lodash.truncate'
import { parseURL, resolveURL, withHttps, withoutTrailingSlash } from 'ufo'
import type { ArticleMeta, AuthorMeta, DeskMeta, ResourcePage, Resources, TagMeta } from '../types'
import { invalidContext } from '../utils/invalid-context'
import type { getSite, useArticleFilter, useHead, useRuntimeConfig, useSeoMeta } from '#imports'

interface SEOItem {
  title: string
  description: string
}

interface SEOConfig {
  meta: SEOItem
  og: SEOItem
  ogImage: string
}

export interface SEOInput extends SEOItem {
  plaintext: string
  headline: string
  cover: { url: string }
  seo: SEOConfig
}

export type RawSEOInput = PartialDeep<SEOInput> & Record<string, any>

function path(object: RawSEOInput, p: string[]): string | undefined {
  return pathOr(object as Record<string, any>, p as any, undefined)
}

function createFirstFound(paths: string[][]) {
  return (input: RawSEOInput): string | undefined => {
    return pipe(
      paths,
      map((p) => path(input, p)),
      filter(isTruthy),
      filter<string | undefined>(Boolean),
      first(),
    )
  }
}

const TITLE = [['seo', 'meta', 'title'], ['title'], ['name']]
const DESK_DESCRIPTION = [['deskSEO', 'meta', 'description']]
const DESCRIPTION = [['seo', 'meta', 'description'], ['blurb'], ['plaintext'], ...DESK_DESCRIPTION]
const OG_TITLE = [['seo', 'og', 'title'], ...TITLE]
const OG_DESK_DESCRIPTION = [['deskSEO', 'og', 'description']]
const OG_DESCRIPTION = [['seo', 'og', 'description'], ...OG_DESK_DESCRIPTION, ...DESCRIPTION]
const OG_IMAGE = [['seo', 'ogImage'], ['headline'], ['cover', 'url'], ['avatar']]
const AUTHOR_BIO = [['bio']]
const TYPE_NAME = [['__typename']]

export interface SEOContext {
  metaType?: Resources
  site: Awaited<ReturnType<typeof getSite>>
  runtimeConfig: ReturnType<typeof useRuntimeConfig>
  articleFilter: ReturnType<typeof useArticleFilter>
  useHead: typeof useHead
  useSeoMeta: typeof useSeoMeta
  resourceUrls: {
    // @ts-expect-error unknown
    article: ResourcePage<ArticleMeta>
    desk: ResourcePage<DeskMeta>
    author: ResourcePage<AuthorMeta>
    tag: ResourcePage<TagMeta>
    [key: string]: ResourcePage<TagMeta>
  }
}

type MetaInput = MetaFlatInput & { title?: string }

type SEOHandler<T = MetaInput> = (input: RawSEOInput, context: SEOContext) => T | undefined | false
type SEOPreset<T = MetaInput> = SEOHandler<T> | SEOHandler<T>[]
type SEOMapResult = MetaInput | undefined | false
type SEOMapFn<T> = (input: T, context: SEOContext) => SEOMapResult
type SEOFilterFn<T> = (input: T, ctx: SEOContext) => T

function createSEO<T>(
  pick: (input: RawSEOInput, context: SEOContext) => T,
  map: SEOMapFn<T>,
  filter: SEOFilterFn<T> = identity,
): SEOHandler {
  return (input: RawSEOInput, context: SEOContext) => {
    return map(filter(pick(input, context), context), context)
  }
}

function simpleSEO(
  paths: string[][],
  map: SEOMapFn<string | undefined>,
  filter: SEOFilterFn<string | undefined> = identity,
): SEOHandler {
  return createSEO(createFirstFound(paths), map, filter)
}

function seoHtmlFilter(input: string | undefined, ctx: SEOContext) {
  if (!input) {
    return input
  }

  return ctx.articleFilter(input)
}

interface MetaDefineSEOInput {
  type?: 'meta'
  setup: (options: Record<string, any>) => SEOPreset
}

type DefineSEOInput = MetaDefineSEOInput

type NormalizedSEOHandler = (input: RawSEOInput, context: SEOContext) => void
type NormalizedSEOPreset = (options: Record<string, any>) => NormalizedSEOHandler

function useMeta(seo: MetaInput, ctx: SEOContext) {
  const { title, ...meta } = seo

  if (title) {
    ctx.useHead({ title }, { mode: 'server' })
  }

  ctx.useSeoMeta(meta)
}

interface MetaDefineSEOHandlerInput {
  type?: 'meta'
  handler: SEOHandler
}

type DefineSEOHandlerInput = MetaDefineSEOHandlerInput

export function defineSEOHandler(inputOrHandler: DefineSEOHandlerInput | SEOHandler): NormalizedSEOHandler {
  const { handler } = typeof inputOrHandler === 'function' ? { handler: inputOrHandler } : inputOrHandler

  return (input: RawSEOInput, ctx: SEOContext) => {
    const seo = handler(input, ctx)

    if (seo) {
      useMeta(seo, ctx)
    }
  }
}

export function defineSEOPreset(
  inputOrSetup: DefineSEOInput | ((options: Record<string, any>) => SEOPreset<MetaInput>),
): NormalizedSEOPreset {
  const { setup } = typeof inputOrSetup === 'function' ? { setup: inputOrSetup } : inputOrSetup

  return (options: Record<string, any>) => {
    const maybeHandlers = setup(options)
    const handlers = Array.isArray(maybeHandlers) ? maybeHandlers : [maybeHandlers]
    return (input: RawSEOInput, context: SEOContext) => {
      for (const handle of handlers) {
        const seo = handle(input, context)

        if (seo) {
          useMeta(seo, context)
        }
      }
    }
  }
}

type ResourceType = 'Article' | 'Desk' | 'Tag' | 'User'
const typeMap: Record<ResourceType, Resources> = {
  Article: 'article',
  Desk: 'desk',
  User: 'author',
  Tag: 'tag',
}
function getResourceURL(input: RawSEOInput, context: SEOContext): string | undefined {
  // skipcq: JS-W1043
  const typeName: ResourceType = input.__typename || '_'
  const resourceType = context.metaType || typeMap[typeName]
  const resourceUrls = context.resourceUrls[resourceType]
  if (!resourceUrls?.enable) return undefined

  // skipcq: JS-W1043
  const siteUrl = (context.runtimeConfig?.public?.siteUrl as string) || '/'
  const url = resourceUrls.toURL(input as any, resourceUrls._context ?? invalidContext)
  return withoutTrailingSlash(resolveURL(siteUrl, url))
}

function getTwitterSite(_input: RawSEOInput, context: SEOContext) {
  const twitterLink = context.site?.socials?.Twitter
  if (!twitterLink) return undefined

  const { pathname } = parseURL(withHttps(twitterLink))
  const accountPath = pathname.split('/')[1]
  return accountPath ? `@${accountPath}` : undefined
}

export const basic = defineSEOPreset(({ twitterCard = 'summary_large_image' }) => [
  // Author
  simpleSEO(AUTHOR_BIO, (authorBio) => {
    const bio = truncate(authorBio ?? '', {
      length: 150,
      separator: /,? +/,
    })
    return isDefined(authorBio) && { description: bio, ogDescription: bio, twitterDescription: bio }
  }),

  // Resource
  simpleSEO(TITLE, (title: string | undefined) => isDefined(title) && { title }, seoHtmlFilter),
  simpleSEO(OG_TITLE, (ogTitle) => isDefined(ogTitle) && { ogTitle, twitterTitle: ogTitle }, seoHtmlFilter),
  simpleSEO(DESCRIPTION, (description) => isDefined(description) && { description }, seoHtmlFilter),
  simpleSEO(
    OG_DESCRIPTION,
    (ogDescription) => isDefined(ogDescription) && { ogDescription, twitterDescription: ogDescription },
    seoHtmlFilter,
  ),
  simpleSEO(OG_IMAGE, (ogImage) => isDefined(ogImage) && { ogImage, twitterImage: ogImage }),

  // Common
  simpleSEO(TYPE_NAME, (typeName) => {
    const type = typeName as ResourceType
    if (type === 'Article') return { ogType: 'article' }
    return { ogType: 'website' }
  }),
  createSEO(getResourceURL, (ogUrl) => isDefined(ogUrl) && { ogUrl }),
  createSEO(getTwitterSite, (twitterSite) => isDefined(twitterSite) && { twitterSite }),
  () => ({ twitterCard }),
])

const emptyPreset = defineSEOPreset(() => [])

const presets: Record<string, NormalizedSEOPreset> = {
  basic,
  __empty: emptyPreset,
}

export const builtinPresets = new Set(Object.keys(presets))

export interface PresetConfig {
  preset?: string
  presetFactory?: NormalizedSEOPreset
  options?: Record<string, any>
}

type InlineSEOPreset = [preset: NormalizedSEOPreset, options?: Record<string, any>]

export type PresetConfigInput = PresetConfig | InlineSEOPreset | NormalizedSEOHandler

export function resolveSEOPresets(configs: PresetConfigInput[]): NormalizedSEOHandler[] {
  return configs.map((config: PresetConfigInput) => {
    if (typeof config === 'function') {
      return config
    } else if (Array.isArray(config)) {
      const [presetFactory, options = {}] = config
      return presetFactory(options)
    }

    if (config.presetFactory) {
      return config.presetFactory(config.options || {})
    }

    return (presets[config.preset || '__empty'] || emptyPreset)(config.options || {})
  })
}
