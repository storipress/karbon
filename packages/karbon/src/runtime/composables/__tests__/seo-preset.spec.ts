import { expect, it } from 'vitest'
import { identity, reduce } from 'remeda'
import type { SEOContext } from '../seo-preset'
import { resolveSEOPresets } from '../seo-preset'
import type { BaseMeta, ResourcePage } from '../../types'

it('can handle article', () => {
  const handlers = resolveSEOPresets([{ preset: 'basic' }])
  const results: unknown[] = []
  const resourceURL: ResourcePage<BaseMeta> = {
    enable: true,
    isValid: () => true,
    getIdentity: () => ({ type: 'article', id: '1' }),
    toURL: () => '',
    route: '',
  }
  const context: SEOContext = {
    metaType: 'article',
    runtimeConfig: {} as any,
    site: {
      name: 'site_name',
    },
    articleFilter: identity,
    useHead: (input: unknown) => void results.push(input),
    useSeoMeta: (input: unknown) => void results.push(input),
    resourceUrls: {
      article: {
        ...resourceURL,
        toURL: () => 'article_url',
      },
      desk: {
        ...resourceURL,
        toURL: () => 'desk_url',
      },
      author: {
        ...resourceURL,
        toURL: () => 'author_url',
      },
      tag: {
        ...resourceURL,
        toURL: () => 'tag_url',
      },
    },
  }

  const article = {
    title: 'title',
    blurb: 'blurb',
    bio: 'author_bio',
    cover: {
      url: 'cover_url',
    },
    author: {
      name: 'author_name',
    },
  }

  for (const handler of handlers) {
    handler(article, context)
  }

  expect(reduce(results, (acc, cur) => Object.assign(acc, cur), {})).toMatchSnapshot()
})

const ARTICLE_FIXTURE = {
  __typename: 'Article',
  slug: 'article_slug',
  featured: false,
  layout: null,
  shadow_authors: null,
  metafields: [],
  relevances: [
    {
      __typename: 'Article',
      id: '2',
      title: '<p>Another Article</p>',
    },
  ],
  content_blocks: [],
  bio: 'My bio',
  bioHTML: '<p>My bio</p>',
  published_at: '2024-01-01T00:00:00+00:00',
  updated_at: '2024-01-01T00:00:00+00:00',
  title: '<p>title</p>',
  blurb: '<p>blurb</p>',
  seo: {
    og: {
      title: 'og_title',
      description: 'og_description',
    },
    meta: {
      title: 'meta_title',
      description: 'meta_description',
    },
    hasSlug: true,
    ogImage: '',
  },
  plaintext: 'Hello world',
  cover: {
    alt: 'cover alt',
    url: 'https://example.com/cover.png',
    crop: {
      key: '123',
      top: 50,
      left: 50,
      zoom: 1,
      width: 0,
      height: 0,
      realWidth: 1600,
      realHeight: 915,
    },
    caption: '',
  },
  authors: [
    {
      __typename: 'User',
      bio: '<p>My bio</p>',
      slug: 'author_slug',
      avatar: 'https://example.com/profile.png',
      email: 'author@example.com',
      location: 'Earth',
      first_name: 'Author',
      last_name: 'Name',
      full_name: 'Author Name',
      id: '1',
      socials: {
        LinkedIn: 'www.example.com',
      },
      name: 'Author Name',
    },
  ],
  desk: {
    __typename: 'Desk',
    id: '5',
    name: 'Desk',
    slug: 'desk',
    layout: null,
    desk: null,
  },
  tags: [],
  id: '1',
  plan: 'free',
  html: '<p>Hello world</p>',
  segments: [
    {
      id: 'normal',
      type: 'p',
      html: '<p>Hello world</p>',
    },
  ],
  __sp_cf: {},
  __sp_cf_editor_block: {},
}

it('can handle real article', () => {
  const handlers = resolveSEOPresets([{ preset: 'basic' }])
  const results: unknown[] = []
  const resourceURL: ResourcePage<BaseMeta> = {
    enable: true,
    isValid: () => true,
    getIdentity: () => ({ type: 'article', id: '1' }),
    toURL: () => '',
    route: '',
  }
  const context: SEOContext = {
    metaType: 'article',
    runtimeConfig: {} as any,
    site: {
      name: 'site_name',
    },
    articleFilter: identity,
    useHead: (input: unknown) => void results.push(input),
    useSeoMeta: (input: unknown) => void results.push(input),
    resourceUrls: {
      article: {
        ...resourceURL,
        toURL: () => 'article_url',
      },
      desk: {
        ...resourceURL,
        toURL: () => 'desk_url',
      },
      author: {
        ...resourceURL,
        toURL: () => 'author_url',
      },
      tag: {
        ...resourceURL,
        toURL: () => 'tag_url',
      },
    },
  }

  for (const handler of handlers) {
    handler(ARTICLE_FIXTURE, context)
  }

  expect(reduce(results, (acc, cur) => Object.assign(acc, cur), {})).toMatchSnapshot()
})

const USER_FIXTURE = {
  __typename: 'User',
  id: '16',
  slug: 'user_slug',
  email: 'user@example.com',
  first_name: 'User',
  last_name: 'Name',
  full_name: 'User Name',
  avatar: 'https://example.com/profile.png',
  location: 'Earth',
  bio: '<p>Hello world</p>',
  website: 'example.com',
  socials: {
    Twitter: 'twitter.example.com',
    Facebook: 'fb.example.com',
    LinkedIn: 'linkedin.example.com',
    YouTube: 'www.youtube.com/example',
    Pinterest: 'www.example.com/a',
  },
  created_at: '2024-01-01T00:00:00+00:00',
  updated_at: '2024-01-01T00:00:00+00:00',
  desks: [],
  name: 'User Name',
  __sp_cf: {},
  __sp_cf_editor_block: {},
}

it('can handle author page', () => {
  const handlers = resolveSEOPresets([{ preset: 'basic' }])
  const results: unknown[] = []
  const resourceURL: ResourcePage<BaseMeta> = {
    enable: true,
    isValid: () => true,
    getIdentity: () => ({ type: 'article', id: '1' }),
    toURL: () => '',
    route: '',
  }
  const context: SEOContext = {
    metaType: 'author',
    runtimeConfig: {} as any,
    site: {
      name: 'site_name',
    },
    articleFilter: identity,
    useHead: (input: unknown) => void results.push(input),
    useSeoMeta: (input: unknown) => void results.push(input),
    resourceUrls: {
      article: {
        ...resourceURL,
        toURL: () => 'article_url',
      },
      desk: {
        ...resourceURL,
        toURL: () => 'desk_url',
      },
      author: {
        ...resourceURL,
        toURL: () => 'author_url',
      },
      tag: {
        ...resourceURL,
        toURL: () => 'tag_url',
      },
    },
  }

  for (const handler of handlers) {
    handler(USER_FIXTURE, context)
  }

  expect(reduce(results, (acc, cur) => Object.assign(acc, cur), {})).toMatchSnapshot()
})
