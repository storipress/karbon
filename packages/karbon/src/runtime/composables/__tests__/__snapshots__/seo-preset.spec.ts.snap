// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`can handle article 1`] = `
{
  "description": "blurb",
  "ogDescription": "blurb",
  "ogImage": "cover_url",
  "ogTitle": "title",
  "ogType": "website",
  "ogUrl": "/article_url",
  "title": "title",
  "twitterCard": "summary_large_image",
  "twitterDescription": "blurb",
  "twitterImage": "cover_url",
  "twitterTitle": "title",
}
`;

exports[`can handle author page 1`] = `
{
  "description": "<p>Hello world</p>",
  "ogDescription": "<p>Hello world</p>",
  "ogImage": "https://example.com/profile.png",
  "ogTitle": "User Name",
  "ogType": "website",
  "ogUrl": "/author_url",
  "title": "User Name",
  "twitterCard": "summary_large_image",
  "twitterDescription": "<p>Hello world</p>",
  "twitterImage": "https://example.com/profile.png",
  "twitterTitle": "User Name",
}
`;

exports[`can handle real article 1`] = `
{
  "description": "meta_description",
  "ogDescription": "og_description",
  "ogImage": "https://example.com/cover.png",
  "ogTitle": "og_title",
  "ogType": "article",
  "ogUrl": "/article_url",
  "title": "meta_title",
  "twitterCard": "summary_large_image",
  "twitterDescription": "og_description",
  "twitterImage": "https://example.com/cover.png",
  "twitterTitle": "og_title",
}
`;
